UnitySDK/Assets/ML-Agents/Scripts/Agent.cs

```diff
@@ -621,7 +640,7 @@ void SendInfoToBrain()
             m_Info.maxStepReached = m_MaxStepReached;
             m_Info.id = m_Id;
 
-            brain.SubscribeAgentForDecision(this);
+            m_Brain.RequestDecision(this);
```

Comment:
would be great if the agent only had to call `RequestDecision` (much better name than `SubscribeAgentForDecision` :joy:) and `DecideAction`.

Reply:
Isn't that what is going on in this PR? 
Is there something I am missing to get there ?

Reply:
Sorry, my comment lacked context.  I like the fact that this is happening here.  I feel like this is the _only_ interaction Agent should have with a brain IMHO.  I don't think Agent should be instantiating Brains itself.

Reply:
Interesting. I don't see how this would happen, since Agent and brains would be tightly coupled in this PR, I was thinking we should have the UI to create the brain live on the Agent inspector. The creation of the factory was to enable the creation of a brain anywhere so I assumed creating in the Agent. If something must pass the Brain to the Agent, we need to think of what that something is and how the user will parameterize it.

Reply:
Not saying this is a requirement for this PR to get approved, but there are a couple of red flags that pop out to me.  Even though the factory 'hides' different classes of the IPolicy implementation, the remnants of their inner workings still remain.
Agents still know about: 
- NNModels
- Heuristic functions
- Inference Device
- Whether or not the communicator is connected
- Whether or not to use a Heuristic brain (which is like a meta leak)

This is all information being leaked into the agent, when in theory, it could just care about the IPolicy it is interacting with and nothing else.  

In terms of ways to get this done, it may be controversial to say, but if IPolicy's were components (MonoBehaviours) you could easily parameterize them and instantiate them without having a factory or any of this information in Agent.  Agent could then call `GetComponent<IPolicy>():` to retrieve the concrete Policy without knowing any of the details of its implementation.  I know we've talked about moving away from Unity Dependencies, but to be honest, deriving from MonoBehaviour and being able to add it as a component feels better to me.